flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunctionBase.java

```diff
@@ -0,0 +1,219 @@
+            case MAP:
+                final MapType mapType = (MapType) argType;
+                return createMapUnnestFunction(
+                        context,
+                        RowType.of(false, mapType.getKeyType(), mapType.getValueType()),
+                        ArrayData.createElementGetter(mapType.getKeyType()),
+                        ArrayData.createElementGetter(mapType.getValueType()));
+            default:
+                throw new UnsupportedOperationException("Unsupported type for UNNEST: " + argType);
+        }
+    }
+
+    protected abstract UserDefinedFunction createCollectionUnnestFunction(
+            SpecializedContext context,
+            LogicalType elementType,
+            ArrayData.ElementGetter elementGetter);
+
+    protected abstract UserDefinedFunction createMapUnnestFunction(
+            SpecializedContext context,
+            RowType keyValTypes,
+            ArrayData.ElementGetter keyGetter,
+            ArrayData.ElementGetter valueGetter);
+
+    public static LogicalType getUnnestedType(LogicalType logicalType, boolean withOrdinality) {
+        LogicalType elementType;
+        switch (logicalType.getTypeRoot()) {
+            case ARRAY:
+                elementType = ((ArrayType) logicalType).getElementType();
+                break;
+            case MULTISET:
+                elementType = ((MultisetType) logicalType).getElementType();
+                break;
+            case MAP:
+                MapType mapType = (MapType) logicalType;
+                elementType = RowType.of(false, mapType.getKeyType(), mapType.getValueType());
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported UNNEST type: " + logicalType);
+        }
+
+        if (withOrdinality) {
+            return wrapWithOrdinality(elementType);
+        }
+        return elementType;
+    }
+
+    public static LogicalType wrapWithOrdinality(LogicalType baseType) {
+        // If baseType is already a ROW, extract its fields and add an ordinality field
+        if (baseType instanceof RowType) {
+            RowType rowType = (RowType) baseType;
+            return new RowType(
+                    false,
+                    Stream.concat(
+                                    rowType.getFields().stream(),
+                                    Stream.of(
+                                            new RowType.RowField(
+                                                    "ordinality",
+                                                    DataTypes.INT().notNull().getLogicalType())))
+                            .collect(Collectors.toList()));
+        } else {
+            // For non-row types, wrap in a row with f0 and ordinality
+            return RowType.of(
+                    false,
+                    new LogicalType[] {baseType, DataTypes.INT().notNull().getLogicalType()},
+                    new String[] {"f0", "ordinality"});
```

Comment:
For PTF [we changed the default](https://github.com/apache/flink/blob/master/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/SystemTypeInference.java#L272) from `f0` to `EXPR$0`. We should do the same here.